### PR TITLE
Add `--tallies` option to `crystal tool unreachable`

### PR DESCRIPTION
**Pull request by straight-shoota**
This option changes `crystal tool unreachable` to output *all* defs and their call counts. The call count of unreachable defs is zero. This additional information enables collecting data across multiple entrypoints with different dependency graphs.

Consider two files serving as different entry points, `entry-a.cr` and `entry-b.cr`. They require some of the same file dependencies, but also different ones. The standard behaviour of `unreachable` can only tell which defs are unreachable in `a` and which are unreachable in `b`. Shared entries are definitely unreachable in both. A def `::foo` that is only listed as unreachable in `a`, not in `b` can mean it's reachable in `b`. But it can also mean `::foo` is not even in the dependency graph of `entry-b.cr` and thus never seen by the compiler.
The `--tallies` option printing *all* defs, reachable and unreachable, resolves this uncertainty by clearly stating whether `::foo` is present in `b` or not.
For this use case, the exact call count wouldn't be necessary. The information whether there's at least one call would suffice. But giving a total number isn't much extra effort and can be useful for further statistical analysis.

The implementation overall is pretty simple, the main change is from `Set` to register presence to a `Hash` for counting occurences.

A demonstration of applying this features on a code base is available at https://gist.github.com/straight-shoota/f7c9120b3bdbc067b4877f88def9c51e
It analyzes the source code of the compiler and standard library using three entry points (compiler, compiler_spec and std_spec) and combines the information into a single database.

### CODE REVIEW

File: src/compiler/crystal/tools/unreachable.cr [hunk 1]
```diff
@@ -199,11 +208,10 @@ module Crystal
...
       check_def(previous) if previous && !a_def.calls_previous_def?
 
-      return if @used_def_locations.includes?(a_def.location)
-
-      check_def(previous) if previous && a_def.calls_previous_def?
+      tally = @used_def_locations[a_def.location]
+      @tallies[a_def] = tally
 
-      @defs << a_def
+      check_def(previous) if previous && a_def.calls_previous_def? if tally == 0
```
Comment (HertzDevil) [inline src/compiler/crystal/tools/unreachable.cr:214]: Are these the same?

```suggestion
      check_def(previous) if previous && a_def.calls_previous_def? && tally == 0
```

Comment (straight-shoota) [reply src/compiler/crystal/tools/unreachable.cr:214]: Yeah, should be 😅

**Review by HertzDevil — APPROVED** [review]

File: src/compiler/crystal/tools/unreachable.cr (hunk 1 above)
Comment (straight-shoota) [reply src/compiler/crystal/tools/unreachable.cr]: I mean the order of execution changes, but there are no side effects on these.

File: src/compiler/crystal/tools/unreachable.cr [hunk 2]
```diff
@@ -199,11 +208,10 @@ module Crystal
...
       check_def(previous) if previous && !a_def.calls_previous_def?
 
-      return if @used_def_locations.includes?(a_def.location)
-
-      check_def(previous) if previous && a_def.calls_previous_def?
+      tally = @used_def_locations[a_def.location]
+      @tallies[a_def] = tally
 
-      @defs << a_def
+      check_def(previous) if previous && a_def.calls_previous_def? && tally == 0
```
Comment (Sija) [inline src/compiler/crystal/tools/unreachable.cr:214]: Nitpick:

```suggestion
      check_def(previous) if previous && a_def.calls_previous_def? && tally.zero?
```

**Pull request merged**